### ISSUE TITLE
[DOCS] Migrating from grains.get_or_set_hash to sdb.get_or_set_hash

### ISSUE THREAD

**Issue by hugo-ricateau** [saltstack/salt#62816]
**Description**
According to the [3005 release note](https://docs.saltproject.io/en/latest/topics/releases/3005.html#removed), `grains.get_or_set_hash` has been removed in favour of `sdb.get_or_set_hash` (#61691).

I've updated every occurrences of `salt['grains.get_or_set_hash'](...)` to `salt['sdb.get_or_set_hash'](...)` in the state tree. When applying it via `sudo salt-call --local state.apply` (note that I'm using `salt` in [masterless mode](https://docs.saltproject.io/en/latest/topics/tutorials/quickstart.html)), in place of the generated hash I had before, I'm now obtaining a `False` value.

Even the example from the [documentation](https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.sdb.html#salt.modules.sdb.get_or_set_hash) is failing to generate and return a value:
```
sudo salt-call --local sdb.get_or_set_hash 'SECRET_KEY' 50
local:
    False
```

I've tried to setup [another database](https://docs.saltproject.io/en/latest/topics/sdb/index.html#sdb-configuration) but none of the [existing drivers](https://docs.saltproject.io/en/latest/ref/sdb/all/index.html#sdb-modules) seems to implement the `.get_or_set_hash` method.

Another question: previously, I was sharing the grains file across the nodes of the cluster trough a `glusterfs` mount point in order to share data between masterless minions (master-minion model is not suitable in this specific situation); when using default `sdb` database (documentation is unclear if a default database even exists or if one must necessarily define one with a specific driver), where is the data-storage located?

**Comment by OrangeDog** [comment]
I already fixed that specific example in #62627.

Alternatively, put `grains.get_or_set_hash` back because it's not inherently insecure.

The get-or-set logic is a bit clunky to do in jinja, but that is also an option.

**Comment by hugo-ricateau** [comment]
Thanks for your reply @OrangeDog.

> I already fixed that specific example in #62627.

Okey, so let's analyse
```
% sudo salt-call --local sdb.get_or_set_hash sdb://mymemcached/SECRET_KEY 50
local:
    sdb://mymemcached/SECRET_KEY
```
As far as I'm concerned, there are two confusing aspects on this specific example:
* Why is this returning the name of the key instead of the generated hash? In templating one needs the generated hash, not the querying key.
* What is `mymemcached`, and where it is defined? As I understand it, I should [setup a database](https://docs.saltproject.io/en/latest/topics/sdb/index.html#sdb-configuration) in minion configuration file (@`/etc/salt/minion`); however, as stated in the initial report, no [available driver](https://docs.saltproject.io/en/latest/ref/sdb/all/index.html#sdb-modules) seems to implement the `.get_or_set_hash` method.

> Alternatively, put `grains.get_or_set_hash` back because it's not inherently insecure. Somebody just panicked because of the "trusting grains for key acceptance / pillar mapping" issue and deleted it.

In which version is this supposed to be available again? As of `3005.1` it still fails with
```
% sudo salt-call --local grains.get_or_set_hash 'SECRET_KEY' 50
'grains.get_or_set_hash' is not available.
```

> The get-or-set logic is a bit clunky to do in jinja, but that is also an option.

Could you provide a ressource or an exemple of what you think of?

**Comment by hugo-ricateau** [comment]
Do I need to open specific bug issues or is this one sufficient?

**Comment by lkubb** [comment]
>     * [...] however, as stated in the initial report, no [available driver](https://docs.saltproject.io/en/latest/ref/sdb/all/index.html#sdb-modules) seems to implement the `.get_or_set_hash` method.

It is sufficient for the drivers to implement get and set, the get_or_set_hash logic is implemented [in the utils module](https://github.com/saltstack/salt/blob/501e2324720f1336c27390ee2949d1448c59e024/salt/utils/sdb.py#L122), which is used by the sdb execution module.

**Comment by hugo-ricateau** [comment]
> > * [...] however, as stated in the initial report, no [available driver](https://docs.saltproject.io/en/latest/ref/sdb/all/index.html#sdb-modules) seems to implement the `.get_or_set_hash` method.
> 
> It is sufficient for the drivers to implement get and set, the get_or_set_hash logic is implemented [in the utils module](https://github.com/saltstack/salt/blob/501e2324720f1336c27390ee2949d1448c59e024/salt/utils/sdb.py#L122), which is used by the sdb execution module.

I agree with you @lkubb that, in theory, it should; however, in practice, it does not seem to be the case: I tried calling `sdb.get_or_set_hash` against a database defined with a [`yaml`](https://docs.saltproject.io/en/latest/ref/sdb/all/salt.sdb.yaml.html#module-salt.sdb.yaml) driver, and it resulted in an uncaught `NotImplementedError` on `salt-call`.

**Comment by lkubb** [comment]
>[...] I tried calling `sdb.get_or_set_hash` against a database defined with a [`yaml`](https://docs.saltproject.io/en/latest/ref/sdb/all/salt.sdb.yaml.html#module-salt.sdb.yaml) driver, and it resulted in an uncaught `NotImplementedError` on `salt-call`.

That's because the YAML sdb module does not implement ``set_``: https://github.com/saltstack/salt/blob/501e2324720f1336c27390ee2949d1448c59e024/salt/sdb/yaml.py#L63-L67

**Comment by hugo-ricateau** [comment]
> That's because the YAML sdb module does not implement `set_`

I was expecting such a reason; however, my point here is that such an implementation defeats the purpose of a generic db interface: requiring manual edits in place of automatic generation is a non-sens for a database interface.

In summary, I think there are several issues reported in this thread:
* `grains.get_or_set_hash` should be restored, and a post-mortem should conclude with the necessity of adding safeguards to prevent such breaking changes to be so casually merged in the future.
* All sdb drivers should fully implement the interface (or being tagged as not fully functional).
* The documentation concerning the sdb module should be clarified: [it](https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.sdb.html#module-salt.modules.sdb) harshly misses a generic description of how the backend works, and especially of a link to the documentation describing [how to define a sdb database](https://docs.saltproject.io/en/latest/topics/sdb/index.html#storing-data-in-other-databases); then, this later page also misses a link to the list of [available drivers](https://docs.saltproject.io/en/latest/ref/sdb/all/index.html#sdb-modules), along with some explanation of what has been pointed-out in this thread.

**Comment by barbaricyawps** [comment]
Can either @OrangeDog or @hugo-ricateau provide some feedback about what action needs to be taken here from a docs perspective? I'll leave the `needs-triage` label on so that I take a look at this again next week.

**Comment by barbaricyawps** [comment]
I think that this is bigger than a docs issue. I'm going to remove the Docs label so that it will get picked up by the developer's triage process.

**Comment by vincent-olivert-riera** [comment]
Any update on this?

**Comment by jbondhus** [comment]
Is there any update on this? This issue is still happening on salt 3007, it returns the name of the key instead of the value. I can't use grains.get_or_set_hash either because it's been removed.

<img width="999" alt="Screenshot 2024-04-13 at 11 37 33 PM" src="https://github.com/saltstack/salt/assets/1085350/736fcafb-fe0f-4b19-8ed5-6261a3398af2">
